servers/src/main/java/tachyon/worker/block/evictor/GreedyEvictor.java

```diff
@@ -0,0 +1,160 @@
+    if (location.equals(BlockStoreLocation.anyTier())) {
+      selectedDir = selectDirToEvictBlocksFromAnyTier(availableBytes);
+    } else {
+      int tierAlias = location.tierAlias();
+      StorageTier tier = mMetaManager.getTier(tierAlias);
+      if (location.equals(BlockStoreLocation.anyDirInTier(tierAlias))) {
+        selectedDir = selectDirToEvictBlocksFromTier(tier, availableBytes);
+      } else {
+        int dirIndex = location.dir();
+        StorageDir dir = tier.getDir(dirIndex);
+        if (canEvictBlocksFromDir(dir, availableBytes)) {
+          selectedDir = dir;
+        }
+      }
+    }
+    if (selectedDir == null) {
+      LOG.error("Failed to freeSpace: No StorageDir has enough capacity of {} bytes",
+          availableBytes);
+      return null;
+    }
+
+    // 2. Check if the selected StorageDir already has enough space.
+    List<Pair<Long, BlockStoreLocation>> toTransfer =
+        new ArrayList<Pair<Long, BlockStoreLocation>>();
+    List<Long> toEvict = new ArrayList<Long>();
+    long bytesAvailableInDir = selectedDir.getAvailableBytes();
+    if (bytesAvailableInDir >= availableBytes) {
+      // No need to evict anything, return an eviction plan with empty instructions.
+      return new EvictionPlan(toTransfer, toEvict);
+    }
+
+    // 3. Collect victim blocks from the selected StorageDir. They could either be evicted or
+    // moved.
+    List<BlockMeta> victimBlocks = new ArrayList<BlockMeta>();
+    for (BlockMeta block : selectedDir.getBlocks()) {
+      victimBlocks.add(block);
+      bytesAvailableInDir += block.getBlockSize();
+      if (bytesAvailableInDir >= availableBytes) {
+        break;
+      }
+    }
+
+    // 4. Make best effort to transfer victim blocks to lower tiers rather than evict them.
+    Map<StorageDir, Long> pendingBytesInDir = new HashMap<StorageDir, Long>();
+    for (BlockMeta block : victimBlocks) {
+      StorageTier fromTier = block.getParentDir().getParentTier();
+      List<StorageTier> toTiers = mMetaManager.getTiersBelow(fromTier.getTierAlias());
+      StorageDir toDir = selectDirToTransferBlock(block, toTiers, pendingBytesInDir);
+      if (toDir == null) {
+        // Not possible to transfer
+        toEvict.add(block.getBlockId());
+      } else {
+        StorageTier toTier = toDir.getParentTier();
+        toTransfer.add(new Pair<Long, BlockStoreLocation>(block.getBlockId(),
+            new BlockStoreLocation(toTier.getTierAlias(), toTier.getTierLevel(), toDir
+                .getDirIndex())));
+        if (pendingBytesInDir.containsKey(toDir)) {
+          pendingBytesInDir.put(toDir, pendingBytesInDir.get(toDir) + block.getBlockSize());
+        } else {
+          pendingBytesInDir.put(toDir, block.getBlockSize());
+        }
+      }
+    }
+    return new EvictionPlan(toTransfer, toEvict);
+  }
+
+  private boolean canEvictBlocksFromDir(StorageDir dir, long availableBytes) {
+    return dir.getCapacityBytes() > availableBytes;
```

Comment:
if some blocks are locked in the dir, the locked blocks should not be evicted. and this dir may not be able to free enough space